test/unit/s3gateway_test.js

```diff
@@ -265,7 +267,72 @@ function testEscapeURIPathPreservesDoubleSlashes() {
     }
 }
 
-function test() {
+async function testEcsCredentialRetrieval() {
+    process.env['S3_ACCESS_KEY_ID'] = undefined;
+    process.env['AWS_CONTAINER_CREDENTIALS_ABSOLUTE_URI'] = 'http://localhost/example';
+    globalThis.ngx.fetch = function (url) {
+        globalThis.recordedUrl = url;
+
+        return Promise.resolve({
+            json: function () {
+                return Promise.resolve({
+                    AccessKeyId: 'AN_ACCESS_KEY_ID',
+                    Expiration: '2017-05-17T15:09:54Z',
+                    RoleArn: 'TASK_ROLE_ARN',
+                    SecretAccessKey: 'A_SECRET_ACCESS_KEY',
+                    Token: 'A_SECURITY_TOKEN',
+                });
+            }
+        });
+    };
+
+    await s3gateway.fetchCredentials();
+
+    if (globalThis.recordedUrl !== 'http://localhost/example') {
+        throw 'No or wrong ECS credentials fetch URL recorded: ' + globalThis.recordedUrl;
+    }
+}
+
+async function testEc2CredentialRetrieval() {
+    process.env['S3_ACCESS_KEY_ID'] = undefined;
+    process.env['AWS_CONTAINER_CREDENTIALS_ABSOLUTE_URI'] = undefined;
+    globalThis.ngx.fetch = function (url, options) {
+        if (url === 'http://169.254.169.254/latest/api/token' && options && options.method === 'PUT') {
+            return Promise.resolve({
+                text: function () {
+                    return Promise.resolve('A_TOKEN');
+                },
+            });
+        } else if (url === 'http://169.254.169.254/latest/meta-data/iam/security-credentials/s3access') {
+            if (options && options.headers && options.headers['x-aws-ec2-metadata-token'] === 'A_TOKEN') {
+                return Promise.resolve({
+                    json: function () {
+                        globalThis.credentialsIssues = true;
+                        return Promise.resolve({
+                            AccessKeyId: 'AN_ACCESS_KEY_ID',
+                            Expiration: '2017-05-17T15:09:54Z',
+                            RoleArn: 'TASK_ROLE_ARN',
+                            SecretAccessKey: 'A_SECRET_ACCESS_KEY',
+                            Token: 'A_SECURITY_TOKEN',
+                        });
+                    },
+                });
+            } else {
+                throw 'Invalid token passed: ' + options.headers['x-aws-ec2-metadata-token'];
+            }
+        } else {
+            throw 'Invalid request URL: ' + url;
+        }
+    };
+
+    await s3gateway.fetchCredentials();
```

Comment:
A mock request object is not passed to the `s3gateway.fetchCredentials()` like in other tests. This typically looks like:
```javascript
    var r = {
        "headersOut" : {
            "Accept-Ranges": "bytes",
            "Content-Length": 42,
            "Content-Security-Policy": "block-all-mixed-content",
            "Content-Type": "text/plain",
            "X-Amz-Bucket-Region": "us-east-1",
            "X-Amz-Request-Id": "166539E18A46500A",
            "X-Xss-Protection": "1; mode=block"
        }
    }

    r.log = function(msg) {
        console.log(msg);
    }
```
This causes a unit test failure in this function (which is called s3gateway.js:697):
```javascript
function _debug_log(r, msg) {
    if (debug && "log" in r) {
        r.log(msg);
    }
}
```

Reply:
👍